File: cpp/SortCharactersByFrequence.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <algorithm>


using namespace std;
class Solution {
public:
    string frequencySort(string s) {
        unordered_map<char, int> count;
        string res = "";

        for(int i = 0; i < s.size(); i++){
            count[s[i]] += 1;
        }

        vector<char> sortedC;
        for (const auto& pair : count) {
            sortedC.push_back(pair.first);
        }

        sort(sortedC.begin(), sortedC.end(), [&](char a, char b) {
            return count[a] > count[b];
        });
        for(char c : sortedC){
            res += string(count[c], c);
        }

        return res;
    }
};

File: python/SortCharactersByFrequence.py
class Solution:

    # A bucket sort algorithm, map each char to frequence and then multiply the char by its frequence
    def frequencySort(self, s: str) -> str:
        countMap = {}
        groups = {}
        res = []
        for c in s:
            if c in countMap:
                countMap[c] += 1
            else:
                countMap[c] = 1

        for char, cnt in countMap.items():
            if cnt in groups:
                groups[cnt].append(char)
            else:
                groups[cnt] = [char]
        
        for i in range(len(s), 0, -1):
            if i in groups:
                for c in groups[i]:
                    res.append(c * i)

        return ''.join(res)